### PR TITLE
Feat: Add TF locals support [CFG-1500]

### PR DESCRIPTION
**Pull request by ofekatr**
### What this does
- Adds support for dereferencing local variables during parsing, including:
    - Expressions evaluation during the dereferencing.
    - Correctly evaluate locals with input values references.
    - Correctly evaluate locals with other local values references.

### Background context

We would like to start supporting reading variable values in Terraform files. This is to help on our vision to create better TF support for our product and find more issues that might have been ignored so far because they would be coming from variable values. This PR is for adding support for local values.

### More information

- [CFG-1500](https://snyksec.atlassian.net/browse/CFG-1500)
- [Terraform - Local values](https://www.terraform.io/language/values/locals)

### CODE REVIEW

File: terraform/utils.go [hunk 1]
```diff
@@ -25,6 +25,15 @@ func isValidInputsFile(fileName string) bool {
...
 }
 
+func isValidLocalsFile(fileName string) bool {
+	for _, fileExt := range VALID_TERRAFORM_FILES {
+		if strings.HasSuffix(fileName, fileExt) {
+			return true
+		}
+	}
+	return false
+}
```
Comment (ipapast) [inline terraform/utils.go:35]: isn't this the same as the `isValidTerraformFile` below? is there something different I'm missing?

File: terraform/parser.go [hunk 2]
```diff
@@ -32,14 +32,18 @@ type NewParserParams struct {
 
 type JSON = map[string]interface{}
 
+func NewParserVariables(variables ModuleVariables) ValueMap {
+	return ValueMap{
+		"var":   cty.ObjectVal(variables.inputs),
+		"local": cty.ObjectVal(variables.locals),
+	}
+}
+
```
Comment (ipapast) [inline terraform/parser.go:41]: was this extraction to help getting it used in dereferenceLocals()?

File: terraform/variables.go [hunk 3]
```diff
@@ -88,10 +109,46 @@ func mergeInputs(inputsByFile InputsByFile) ValueMap {
...
 			combinedInputs[input] = value
 		}
 	}
 
 	return combinedInputs
 }
+
+var maxLocalsDerefIterations = 32
+
+func dereferenceLocals(localExprsMap ExpressionMap, inputs ValueMap) ValueMap {
```
Comment (ipapast) [inline terraform/variables.go:123]: when I tried to read this word, I almost choked 😄 why not add a few more characters and fill it out?
```suggestion
func dereferenceLocals(localExpressionMap ExpressionMap, inputs ValueMap) ValueMap {
```

File: terraform/hcl2json_test.go [hunk 4]
```diff
@@ -939,6 +939,182 @@ variable "dummy" {
 				"debugLogs": map[string]interface{}{},
 			},
 		},
+		{
+			name: "Correctly dereferencing local values in a single file",
```
Comment (ipapast) [inline terraform/hcl2json_test.go:943]: These tests are great, but I would still add a test case for having duplicates locals as this was discussed twice and it's not as easy to spot. It doesn't have to be in this PR, and it can be enhanced in a different one.

Tests are also documentation of the expected behaviour so that the next person does not have to go through the entire implementation to spot something like this.

File: terraform/variables.go [hunk 5]
```diff
@@ -124,3 +117,38 @@ func mergeInputs(inputsByFile InputsByFile) ValueMap {
 
 	return combinedInputs
 }
+
+var maxLocalsDerefIterations = 32
```
Comment (ipapast) [inline terraform/variables.go:121]: This reminds me of the nested directories depth chat. Is there any concern on the number of the chain of locals again? any limitation on the reflect package you're using?
If we think that this is not even a common scenario, why limit ourselves? what do we have to gain here?

Comment (ipapast) [reply terraform/variables.go:121]: For example, I prefer getting an error because a user tried to scan something with 23594495 nested locals, than a user having 33 locals and not providing them accurate results.

**Review by ipapast — APPROVED** [review]
I reviewed all but the last commit (locals referring locals):

We had a chat with @ofekatr and @YairZ101 on slack - I am not able to review the last commit today. as I'm going on holidays I suggested either someone else reviewing that one, or approving this PR and moving the last commit in a new one for a different person. whatever works for you best. Pre-approving the 4/5 commits

**Comment by teodora-sandu** [comment]
@ofekatr or @YairZ101 Could you please open a separate PR for the last commit like Ilianna said? Then you can add me as a reviewer. Thank youu

File: terraform/hcl2json_test.go (hunk 4 above)
Comment (YairZ101) [reply terraform/hcl2json_test.go:943]: > test case for having duplicates locals

`terraform validate` will fail in this case, why do we need to test such cases?

File: terraform/variables.go (hunk 3 above)
Comment (YairZ101) [reply terraform/variables.go]: This change was part of the locals that reference other local commit so we will follow up on this review in the other PR.

File: terraform/utils.go (hunk 1 above)
Comment (YairZ101) [reply terraform/utils.go:35]: it is indeed the same :)

File: terraform/parser.go (hunk 2 above)
Comment (YairZ101) [reply terraform/parser.go:41]: The purpose of this extraction is to help us configure the variables map that we are using in our parser.

File: terraform/hcl2json_test.go (hunk 4 above)
Comment (YairZ101) [reply terraform/hcl2json_test.go:943]: I'm going to merge the PR so let's continue the convo in [this thread](https://snyk.slack.com/archives/C030WL9BLSW/p1646575964229079).
